Safari doesn't translate iframe content

rniwa opened the pull request:
#### 09c4983c5bd980390426e4b15d18f32457a0aabc
<pre>
Safari doesn&apos;t translate iframe content
<a href="https://bugs.webkit.org/show_bug.cgi?id=253353">https://bugs.webkit.org/show_bug.cgi?id=253353</a>
&lt;rdar://59693219&gt;

Reviewed by Darin Adler, Sihui Liu and Wenson Hsieh.

Optionally translate contents within each frame when includeSubframes is set in
_WKTextManipulationConfiguration. To do this, each TextManipulationItem and
TextManipulationControllerManipulationFailure are augmented with a frame identifier.
These IDs are serialized with &quot;-&quot; as delimiter for _WKTextManipulationItem.

_WKTextManipulationItem now exposes two new booleans: isSubframe and isCrossSiteSubframe.
These two booleans are ignored when completing items.

WebPage::startTextManipulations now invokes startTextManipulationForFrame on each frame.
We also detect when a new frame is inserted in WebPage::didCommitLoad.

Finally, in WebPage::completeTextManipulation, we aggregate and complete items per frame.

* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::TextManipulationController::addItem): Add the frame identifier.
(WebCore::TextManipulationController::completeManipulation): Check the frame identifier.

* Source/WebCore/editing/TextManipulationControllerManipulationFailure.h:
* Source/WebCore/editing/TextManipulationItem.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _startTextManipulationsWithConfiguration:completion:]):
(coreTextManipulationItemIdentifierFromString):
(-[WKWebView _completeTextManipulation:completion:]):
(wkTextManipulationErrors):
(-[WKWebView _completeTextManipulationForItems:completion:]):

* Source/WebKit/UIProcess/API/Cocoa/_WKTextManipulationConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextManipulationItem.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextManipulationItem.mm:
(-[_WKTextManipulationItem initWithIdentifier:tokens:isSubframe:isCrossSiteSubframe:]): Added.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::startTextManipulations):
(WebKit::WebPage::startTextManipulationForFrame): Extracted out of startTextManipulations.
(WebKit::WebPage::completeTextManipulation): Complete per frame and aggregate the results.

* Source/WebKit/WebProcess/WebPage/WebPage.h:

* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextManipulation.mm:
(TextManipulation.StartTextManipulationDoesNotFindContentInIframeIfIncludeSubframeIsNotSet): Added.
(TextManipulation.StartTextManipulationFindsContentInIframeIfIncludeSubframeIsSet): Added.
(TextManipulation.StartTextManipulationFindsContentInIframeInsertedLater): Added.
(TextManipulation.CompleteTextManipulationReplaceContentInIframe): Added.
(TextManipulation.CompleteTextManipulationFailWhenContentIsChangedInIframe): Added.

Canonical link: <a href="https://commits.webkit.org/261671@main">https://commits.webkit.org/261671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f33cd3084cbd8987f39f2c8b70aae0a171250a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4259 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121057 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116546 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22967 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5400 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118247 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100254 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105536 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/804 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46069 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13976 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/837 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14668 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20007 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52842 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16490 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4454 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2241,7 +2242,13 @@ - (void)_startTextManipulationsWithConfiguration:(_WKTextManipulationConfigurati
+            identifierBuilder.append('-');
+            identifierBuilder.append(String::number(item.frameID.object().toUInt64()));
+            identifierBuilder.append('-');
+            identifierBuilder.append(String::number(item.identifier.toUInt64()));
```
This can be:
```
auto identifier = makeString(item.frameID.processIdentifier(), '-', item.frameID.object(), '-', item.identifier);
```
Because `ObjectIdentifier` has a `StringTypeAdapter`, and `makeString` is more efficient than `StringBuilder`.

rniwa replied:
Fixed.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2255,9 +2262,26 @@ - (void)_startTextManipulationsWithConfiguration:(_WKTextManipulationConfigurati
+static std::optional<ItemIdentifiers> coreTextManipulationItemIdentifierFromString(NSString *identifier)
 {
-    return makeObjectIdentifier<WebCore::TextManipulationItemIdentifierType>(identifier.longLongValue);
+    auto identifiers = String(identifier).split('-');
```
`StringView::split` is more efficient. So we can make a `String`, but then should `StringView::split` it.

rniwa replied:
Fixed.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2255,9 +2262,26 @@ - (void)_startTextManipulationsWithConfiguration:(_WKTextManipulationConfigurati
+
+    return ItemIdentifiers { WebCore::ProcessQualified(makeObjectIdentifier<WebCore::FrameIdentifierType>(*frameID),
+    makeObjectIdentifier<WebCore::ProcessIdentifierType>(*processID)),
+    makeObjectIdentifier<WebCore::TextManipulationItemIdentifierType>(*itemID) };
```
Need to indent more here.

rniwa replied:
Oops, fixed.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2338,7 +2371,10 @@ - (void)_completeTextManipulationForItems:(NSArray<_WKTextManipulationItem *> *)
-        coreItems.uncheckedAppend(WebCore::TextManipulationItem { coreTextManipulationItemIdentifierFromString(wkItem.identifier), WTFMove(coreTokens) });
+        auto identifiers = coreTextManipulationItemIdentifierFromString(wkItem.identifier);
+        if (!identifiers)
+            continue; // FIXME: Generate an error.
```
Why `FIXME` and not do it now?

rniwa replied:
Oops, I totally forgot about this FIXME. I'll just pass on initialized itemID / frameID since TextManipulationController checks the validity of these IDs already.

darinadler commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7929,20 +7941,78 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
-    auto* controller = mainDocument->textManipulationControllerIfExists();
-    if (!controller) {
-        completionHandler(true, { });
+    FrameIdentifier currentFrameID = items[0].frameID;
```
`auto`?

rniwa replied:
Fixed.

darinadler commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7929,20 +7941,78 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
+
+    auto completeManipulationForItems = [&](const Vector<WebCore::TextManipulationItem>& items) -> std::optional<Vector<TextManipulationControllerManipulationFailure>> {
+        ASSERT(!items.isEmpty());
+        WebFrame* frame = WebProcess::singleton().webFrame(currentFrameID);
```
`auto*`?

rniwa replied:
Sure, fixed.

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6895,6 +6895,9 @@ void WebPage::didCommitLoad(WebFrame* frame)
     resetFocusedElementForFrame(frame);
 
+    if (m_textManipulationExclusionRules)
+        startTextManipulationForFrame(*frame->coreFrame());
```
Do we need to ensure text manipulation does not start on the same frame twice with this change? 
Maybe we could assert that `startObservingParagraphs` will not be called twice?

rniwa replied:
Oh, that's good point. I'd add a check to `WebPage::startTextManipulationForFrame`

szewai commented on Source/WebCore/editing/TextManipulationController.cpp:
```diff
@@ -664,15 +665,21 @@ auto TextManipulationController::completeManipulation(const Vector<WebCore::Text
-            failures.append(ManipulationFailure { identifier, i, ManipulationFailure::Type::InvalidItem });
+        auto frameID = itemToComplete.frameID;
+        auto itemID = itemToComplete.identifier;
+        if (!frameID || !m_document || !m_document->frame() || frameID != m_document->frame()->frameID()) {
```
Is it possible that `m_document` or `m_document->frame()` turns from non-null to null during iteration? If not can we just check before the for loop?

rniwa replied:
Yes. In the middle of replace can end up executing arbitrary scripts, which can obviously navigate or remove the iframe.

szewai replied:
Then it seems we should fill `failures` with error and return early (without executing `if (!containersWithoutVisualOverflowBeforeReplacement.isEmpty())`)

rniwa replied:
I don't think that kind of code complexity is warranted given this will be a very rare condition.

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7929,20 +7941,78 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
+    for (auto& item : items) {
+        if (currentFrameID != item.frameID)
+            containsItemsForMultipleFrames = true;
+    }
```
`WTF::anyOf`?

rniwa replied:
Done.

szewai commented on Source/WebCore/editing/TextManipulationController.cpp:
```diff
@@ -682,7 +689,7 @@ auto TextManipulationController::completeManipulation(const Vector<WebCore::Text
         auto failureOrNullopt = replace(itemData, itemToComplete.tokens, containersWithoutVisualOverflowBeforeReplacement);
         if (failureOrNullopt)
-            failures.append(ManipulationFailure { identifier, i, *failureOrNullopt });
+            failures.append(ManipulationFailure { frameID, itemID, i, *failureOrNullopt });
```
Seems can be a one-liner:

```
if (auto failure = replace(itemData, itemToComplete.tokens, containersWithoutVisualOverflowBeforeReplacement))
    failures.append(...)
```

rniwa replied:
Done.

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7929,20 +7941,78 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
+            failuresForAllItems.append(WTFMove(failure));
+            failuresForAllItems.last().index += startIndexForCurrentFrame;
+        }
+    };
```
IIUC we first iterate `items` to check if they are from the same frame, if it is, we invoke `completeManipulation` once and pass `items`; if it is not, we will iterate `items` again and invoke `completeManipulation` multiple times with subarray of `items` (item in the same subarray is from the same frame)? 

If so it seems we could merge this two cases together. Something like (I didn't try locally so may not compile):
```
unsigned itemSize = 0;
Vector<WebCore::TextManipulationController::ManipulationFailure> failures;
for (unsigned index = 0; index < items.size(); ++index) {
    if (index > 0 && items[index].frameID != items[index-1].frameID) {
         Vector<webcore::TextManipulationItem> subItems { items.span().subspan(index - itemSize, itemSize) };
         failures.appendVector(controller->completeManipulation(subItems));
         itemSize = 0;
    }
    ++itemSize;
}

// Maybe pass items to completeManipulation directly if itemSize == items.size()
Vector<webcore::TextManipulationItem> subItems { items.span().subspan(index - itemSize, itemSize) };
failures.appendVector(controller->completeManipulation(subItems));
```

rniwa replied:
We need to adjust the index of each failure when subspan Vector fails in completeManipulation. This is why I have two code paths. We also need to create failure items for when completeManipulationForItems returns nullopt.

rniwa replied:
Having said that, we can use subspan. That's slightly more efficient in allocation.

szewai replied:
Maybe we could pass `offset` to `completeManipulation`? Like:

```
auto createFailuresForItems = [](const Vector<WebCore::TextManipulationItem>& items, unsigned offset) {
    return WTF::map(items, [offset](const auto& item) mutable {
        return TextManipulationControllerManipulationFailure { item.frameID, item.identifier, offset++, TextManipulationControllerManipulationFailure::Type::NotAvailable };
    });
};
auto completeManipulationForItems = [&](const Vector<WebCore::TextManipulationItem>& items, unsigned offset) mutable {
    ASSERT(!items.isEmpty());
    auto* frame = WebProcess::singleton().webFrame(items[0].frameID);
    if (!frame || !frame->coreFrame() || !frame->coreFrame()->document() || !frame->coreFrame()->document()->textManipulationControllerIfExists())
        return createFailuresForItems(items, offset);
    return frame->coreFrame()->document()->textManipulationControllerIfExists()->completeManipulation(items, offset);
};

unsigned itemCount = 0;
Vector<WebCore::TextManipulationController::ManipulationFailure> failures;
for (unsigned index = 0; index <= items.size(); ++index) {
    if (index == items.size() || (index > 0 && items[index].frameID != items[index-1].frameID)) {
        unsigned offset = index - itemCount;
        Vector<webcore::TextManipulationItem> subItems { items.span().subspan(offset, itemCount) };
        failures.appendVector(completeManipulationForItems(subItems, offset));
        itemCount = 0;
    }
    ++itemCount;
}
completionHandler(false, WTFMove(failures));
```

szewai reviewed (commented):
I wonder whether we should add a feature flag to switch between "mainframe only" and "all frames" mode.

szewai commented on Source/WebCore/editing/TextManipulationController.cpp:
```diff
@@ -642,6 +642,9 @@ void TextManipulationController::addItem(ManipulationItemData&& itemData)
     ASSERT(!itemData.tokens.isEmpty());
     auto newID = m_itemIdentifier.generate();
     m_pendingItemsForCallback.append(TextManipulationItem {
+        m_document->frame()->frameID(),
```
Is `frame()` guaranteed to be non-null?

rniwa replied:
Yes.

szewai commented on Source/WebCore/editing/TextManipulationController.cpp:
```diff
@@ -666,25 +669,30 @@ auto TextManipulationController::completeManipulation(const Vector<WebCore::Text
+        auto frameID = itemToComplete.frameID;
+        auto itemID = itemToComplete.identifier;
+        if (!frameID || !m_document || !m_document->frame() || frameID != m_document->frame()->frameID()) {
+            failures.append(ManipulationFailure { frameID, itemID, i, ManipulationFailure::Type::InvalidItem });
```
This should be `Type::NotAvailable`?

rniwa replied:
Fixed

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2204,12 +2205,20 @@ - (void)_startTextManipulationsWithConfiguration:(_WKTextManipulationConfigurati
+        completionHandler();
+        return;
+    }
+    auto mainFrameURL = frame->url();
```
Seems unused

rniwa replied:
Removed.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2255,9 +2265,40 @@ - (void)_startTextManipulationsWithConfiguration:(_WKTextManipulationConfigurati
+    if (!processID || !frameID || !itemID)
+        return std::nullopt;
+
+    return ItemIdentifiers { WebCore::ProcessQualified(makeObjectIdentifier<WebCore::FrameIdentifierType>(*frameID),
+        makeObjectIdentifier<WebCore::ProcessIdentifierType>(*processID)),
+        makeObjectIdentifier<WebCore::TextManipulationItemIdentifierType>(*itemID) };
```
Do we need to validate the ObjectIdentifiers (not 0)?

rniwa replied:
Added checks.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2318,7 +2365,10 @@ - (void)_completeTextManipulation:(_WKTextManipulationItem *)item completion:(vo
-        ASSERT(coreTextManipulationItemIdentifierFromString(item.identifier) == coreFailure.identifier);
+#if ASSERT_ENABLED
+        auto identifiers = coreTextManipulationItemIdentifierFromString(item.identifier);
+        ASSERT(identifiers && identifiers->frameID == coreFailure.frameID && identifiers->itemID == coreFailure.identifier);
```
I think we probably want three asserts, otherwise it seems hard to tell which condition fails

rniwa replied:
Sure, fixed.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2273,15 +2314,19 @@ - (void)_completeTextManipulation:(_WKTextManipulationItem *)item completion:(vo
     Vector<WebCore::TextManipulationItem> coreItems;
     coreItems.reserveInitialCapacity(1);
-    coreItems.uncheckedAppend(WebCore::TextManipulationItem { itemID, WTFMove(tokens) });
+    coreItems.uncheckedAppend(WebCore::TextManipulationItem { identifiers->frameID, false, false, identifiers->itemID, WTFMove(tokens) });
```
I understand we don't need `isSubFrame` info for completeTextManipulation, but it feels a bit weird to convert it to false when we know the real value

rniwa replied:
I don't think this matters. These booleans are ignored for the purpose of completing text manipulations.

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7935,20 +7951,77 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
+    RELEASE_ASSERT(indexForCurrentItem >= itemCount);
+    completeManipulationForCurrentFrame(indexForCurrentItem - itemCount, items.span().subspan(indexForCurrentItem - itemCount, itemCount));
+
+    completionHandler(false, failuresForAllItems);
```
Nit:

```suggestion
    completionHandler(false, WTFMove(failuresForAllItems));
```

rniwa replied:
Fixed

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7935,20 +7951,77 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
+        if (!frame)
+            return std::nullopt;
+
+        auto* coreFrame = frame->coreFrame();
```
Should we use `RefPtr` here to guard the stack variable?

rniwa replied:
Good point. Fixed.

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7935,20 +7951,77 @@ void WebPage::completeTextManipulation(const Vector<WebCore::TextManipulationIte
+
+    auto completeManipulationForItems = [&](const Vector<WebCore::TextManipulationItem>& items) -> std::optional<Vector<TextManipulationControllerManipulationFailure>> {
+        ASSERT(!items.isEmpty());
+        auto* frame = WebProcess::singleton().webFrame(currentFrameID);
```
(Same question about deploying `RefPtr` here)

rniwa replied:
I don't think we need to use RefPtr here since all we're doing with this pointer is to get coreFrame, which is really a trivial function.

szewai reviewed (approved):
r=me

rniwa committed:
Safari doesn't translate iframe content
https://bugs.webkit.org/show_bug.cgi?id=253353
<rdar://59693219>

Reviewed by Darin Adler, Sihui Liu and Wenson Hsieh.

Optionally translate contents within each frame when includeSubframes is set in
_WKTextManipulationConfiguration. To do this, each TextManipulationItem and
TextManipulationControllerManipulationFailure are augmented with a frame identifier.
These IDs are serialized with "-" as delimiter for _WKTextManipulationItem.

_WKTextManipulationItem now exposes two new booleans: isSubframe and isCrossSiteSubframe.
These two booleans are ignored when completing items.

WebPage::startTextManipulations now invokes startTextManipulationForFrame on each frame.
We also detect when a new frame is inserted in WebPage::didCommitLoad.

Finally, in WebPage::completeTextManipulation, we aggregate and complete items per frame.

* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::TextManipulationController::addItem): Add the frame identifier.
(WebCore::TextManipulationController::completeManipulation): Check the frame identifier.

* Source/WebCore/editing/TextManipulationControllerManipulationFailure.h:
* Source/WebCore/editing/TextManipulationItem.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _startTextManipulationsWithConfiguration:completion:]):
(coreTextManipulationItemIdentifierFromString):
(-[WKWebView _completeTextManipulation:completion:]):
(wkTextManipulationErrors):
(-[WKWebView _completeTextManipulationForItems:completion:]):

* Source/WebKit/UIProcess/API/Cocoa/_WKTextManipulationConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextManipulationItem.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextManipulationItem.mm:
(-[_WKTextManipulationItem initWithIdentifier:tokens:isSubframe:isCrossSiteSubframe:]): Added.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::startTextManipulations):
(WebKit::WebPage::startTextManipulationForFrame): Extracted out of startTextManipulations.
(WebKit::WebPage::completeTextManipulation): Complete per frame and aggregate the results.

* Source/WebKit/WebProcess/WebPage/WebPage.h:

* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextManipulation.mm:
(TextManipulation.StartTextManipulationDoesNotFindContentInIframeIfIncludeSubframeIsNotSet): Added.
(TextManipulation.StartTextManipulationFindsContentInIframeIfIncludeSubframeIsSet): Added.
(TextManipulation.StartTextManipulationFindsContentInIframeInsertedLater): Added.
(TextManipulation.CompleteTextManipulationReplaceContentInIframe): Added.
(TextManipulation.CompleteTextManipulationFailWhenContentIsChangedInIframe): Added.

Canonical link: https://commits.webkit.org/261671@main
